cmd/promtool/analyze.go

```diff
@@ -0,0 +1,543 @@
+// Copyright 2023 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"math"
+	"net/http"
+	"net/url"
+	"os"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+
+	"github.com/prometheus/prometheus/model/labels"
+)
+
+var (
+	errNotNativeHistogram = fmt.Errorf("not a native histogram")
+	errNotEnoughData      = fmt.Errorf("not enough data")
+)
+
+type AnalyzeHistogramsConfig struct {
+	metricType string
+	duration   time.Duration
+	end        string
+	step       time.Duration
+	matchers   []string
+}
+
+// run retrieves metrics that look like conventional histograms (i.e. have _bucket
+// suffixes) or native histograms, depending on metricType flag.
+func (c *AnalyzeHistogramsConfig) run(url *url.URL, roundtripper http.RoundTripper) error {
+	if c.metricType != "classichistograms" && c.metricType != "nativehistograms" {
+		return fmt.Errorf("histogram type is %s, must be 'classichistograms' or 'nativehistograms'", c.metricType)
+	}
+
+	ctx := context.Background()
+
+	api, err := newAPI(url, roundtripper, nil)
+	if err != nil {
+		return err
+	}
+
+	var endTime time.Time
+	if c.end != "" {
+		endTime, err = parseTime(c.end)
+		if err != nil {
+			return fmt.Errorf("error parsing end time '%s': %w", c.end, err)
+		}
+	} else {
+		endTime = time.Now()
+	}
+	startTime := endTime.Add(-c.duration)
+
+	if c.metricType == "nativehistograms" {
+		// histoMetrics, err := queryMetadataForHistograms(ctx, api, "", "1000")
+		histoMetrics, err := queryMetricNames(ctx, api, c.matchers, startTime, endTime)
+		if err != nil {
+			return err
+		}
+		return c.getStatsFromMetrics(ctx, api, startTime, endTime, os.Stdout, identity, identity, calcNativeBucketStatistics, histoMetrics)
+	}
+
+	baseMetrics, err := queryBaseMetricNames(ctx, api, startTime, endTime)
```

Comment:
How about we add an option "--matchall" so that the user has to explicitly request finding and analyzing all histograms?

I think there's two use cases:
1. As a user I want to see the impact of moving a specific classic histogram metric to native histograms (filter)
2. As a user I want to see the impact of moving all of my classic histograms to native histograms

Reply:
To note down what we discussed in the histogram sync meeting:

Analysing all histograms would be a nice feature, but it's actually quite hard to find all histograms. For classic histogram, we could just query `{__name__=~".+_bucket"}` and hope that they are all indeed buckets of classic histograms. (We could check that the corresponding `..._count` exists for better confidence.) Native histograms are at least easy to spot in the query response, but to find them all, we essentially have to query _all_ time series. What @zenador tried so far in this PR, namely querying metadata, is also not perfect, because the current metadata store in Prometheus only tells you about current scrapes. Any histogram currently not being scraped would not show up there.

So let's first go with the matcher approach to got something working done which we can merge. In a further step, we can then think about a way to find and analyze all histograms efficiently. (For the expensive way, the user can already specify `{__name__=~".+_bucket"}` for classic histograms or  `{__name__=~".+"}` for native histograms. But this will blow up for large TSDBs.)